File: flask/ms_person.py
'''
web service for handling people records
'''
import json
from flask import Flask, request, jsonify
from person_store import PersonStore


HTTP_NOT_FOUND = 404
HTTP_INVALID   = 403
HTTP_OK        = 200

app = Flask('person')

@app.route('/')
def info():
  return "person microservice (flask)"


@app.route('/help', methods = ['GET'])
def help():
  func_list = {}

  for rule in app.url_map.iter_rules():
    if rule.endpoint != 'static':
      func_list[rule.rule] = app.view_functions[rule.endpoint].__doc__

  return jsonify(func_list), HTTP_OK


@app.route('/people', methods=['GET'])
def people():
  ps     = PersonStore()
  people = ps.people

  return json.dumps(people), HTTP_OK


@app.route('/person/<int:person_id>', methods=['GET'])
def read_person(person_id):
  ps      = PersonStore()
  person  = ps.read_person(person_id) 

  if person is None:
    return "NOT FOUND", HTTP_NOT_FOUND

  return json.dumps(person), HTTP_OK


@app.route('/person/<int:person_id>', methods=['PUT'])
def update_person(person_id):
  person_data = request.json
  ps          = PersonStore()
  person      = ps.update_person(person_id, person_data)

  return jsonify(person), HTTP_OK


@app.route('/person', methods=['POST'])
def create_person():
  person_data = request.json
  ps          = PersonStore()
  person      = ps.create_person(person_data)

  return jsonify(person), HTTP_OK


@app.route('/person/<int:person_id>', methods=['DELETE'])
def delete_person(person_id):
  ps = PersonStore()
  
  if ps.delete_person(person_id):
    return "DELETED", HTTP_OK
  else:
    return "NOT FOUND", HTTP_NOT_FOUND


if __name__ == "__main__":
  app.run(debug=True)

File: sinatra/ms_person.rb
require 'sinatra'
require 'sinatra/advanced_routes'
require './person_store'


HTTP_NOT_FOUND = 404
HTTP_INVALID   = 403
HTTP_OK        = 200


get '/' do 
  status HTTP_OK
  "person microservice (sinatra)"
end


get '/help' do
  Sinatra::Application.each_route do |route|
    puts '-' * 20
    puts route.app.name   # "SomeSinatraApp"
    puts route.path       # that's the path given as argument to get and akin
    puts route.verb       # get / head / post / put / delete
    puts route.file       # "some_sinatra_app.rb" or something
    puts route.line       # the line number of the get/post/... statement
    puts route.pattern    # that's the pattern internally used by sinatra
    puts route.keys       # keys given when route was defined
    puts route.conditions # conditions given when route was defined
    puts route.block      # the route's closure
  end
end


get '/people' do
  status HTTP_OK

  ps     = PersonStore.new()
  people = ps.people

  people.to_json
end


get '/person/:person_id' do
  status HTTP_OK

  person_id = params[:person_id].to_i
  ps        = PersonStore.new()
  person    = ps.read_person(person_id)

  status HTTP_NOT_FOUND if person.nil?

  person.to_json
end


put '/person/:person_id' do
  status HTTP_OK
  request.body.rewind

  person_id   = params[:person_id].to_i
  person_data = JSON.parse request.body.read
  ps          = PersonStore.new()
  person      = ps.update_person(person_id, person_data)

  status HTTP_NOT_FOUND if person.nil?
  person.to_json
end


post '/person' do
  status HTTP_OK
  request.body.rewind

  person_data = JSON.parse request.body.read
  ps          = PersonStore.new()
  person      = ps.create_person(person_data)

  person.to_json
end


delete '/person/:person_id' do
  status HTTP_OK

  person_id = params[:person_id].to_i
  ps        = PersonStore.new()

  status HTTP_NOT_FOUND unless ps.delete_person(person_id)
end

